patches/server/0974-Call-BlockGrowEvent-for-missing-blocks.patch

```diff
@@ -0,0 +1,39 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
+Date: Fri, 9 Jun 2023 13:04:42 +0200
+Subject: [PATCH] Call BlockGrowEvent for missing blocks
+
+Call it for pitcher crops and sniffer egg
+
+diff --git a/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java b/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
+index 849c0ae6ca30691134f51e5d4d5da372e0e728bd..d2e8aafb8db17f2426dd83353c0cfdeaf0d372aa 100644
+--- a/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
+@@ -126,7 +126,7 @@ public class PitcherCropBlock extends DoublePlantBlock implements BonemealableBl
+     private void grow(ServerLevel world, BlockState state, BlockPos pos, int amount) {
+         int i = Math.min(state.getValue(AGE) + amount, 4);
+         if (this.canGrow(world, pos, state, i)) {
+-            world.setBlock(pos, state.setValue(AGE, Integer.valueOf(i)), 2);
++            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(world, pos, state.setValue(AGE, Integer.valueOf(i)), 2)) return; // Paper
+             if (i >= 3) {
+                 BlockPos blockPos = pos.above();
+                 world.setBlock(blockPos, copyWaterloggedFrom(world, pos, this.defaultBlockState().setValue(AGE, Integer.valueOf(i)).setValue(HALF, DoubleBlockHalf.UPPER)), 3);
+diff --git a/src/main/java/net/minecraft/world/level/block/SnifferEggBlock.java b/src/main/java/net/minecraft/world/level/block/SnifferEggBlock.java
+index b8e536c89b00b8f5a853d3b83512d4f0987c08fd..b8a8fd937d36c92a9d9639516b024d98343d5090 100644
+--- a/src/main/java/net/minecraft/world/level/block/SnifferEggBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/SnifferEggBlock.java
+@@ -65,8 +65,13 @@ public class SnifferEggBlock extends Block {
+     @Override
+     public void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+         if (!this.isReadyToHatch(state)) {
++            // Paper start
++            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(world, pos, state.setValue(HATCH, Integer.valueOf(this.getHatchLevel(state) + 1)), 2)) {
++                rescheduleTick(world, pos);
++                return;
++            }
++            // Paper end
+             world.playSound((Player)null, pos, SoundEvents.SNIFFER_EGG_CRACK, SoundSource.BLOCKS, 0.7F, 0.9F + random.nextFloat() * 0.2F);
+-            world.setBlock(pos, state.setValue(HATCH, Integer.valueOf(this.getHatchLevel(state) + 1)), 2);
```

Comment:
So this is more something I'm just noticing. All the other places the BlockGrowEvent is handled, the setBlock call is still there, it just happens after the event is handled. So are you sure we should be removing these setBlock calls? It also results in different update flags being used for the setBlock as the flag param of `handleBlockGrowEvent` isn't used anywhere.

Reply:
Not all the places (if you see the setBlock are not in the same position for the other cases) but yes i'm aware this is something everywhere in the server when the event require a blockstate that is forcibly placed using blockstate#update, the updates are always mangled because the blockstate almost never use the internal flag (same using the BlockStateListPopulator or the bukkit capture system). Owen started a cleanup of the BlockPlaceEvent logic around there for spigot (also add the support to tile entity in the tree). Then i might pr a patch that i wasn't planning to pr that soon but that will restore the right update flag and the right order of updates during the capture of bukkit (i redo the capture system here).

Reply:
Ok, sounds good. I couldn't actually find a difference in behavior between the two flags, but I maybe was just trying the wrong things. Like they both broke nether portals and fired observers.